Text preprocessing is an essential step in natural language processing (NLP) that involves cleaning and transforming unstructured text data to prepare it for analysis. It includes tokenization, stemming, lemmatization, stop-word removal, and part-of-speech tagging. Here we will introduce the basics of text preprocessing and provide Python code examples to illustrate how to implement these tasks using the NLTK library.

### What is Text Preprocessing in NLP?

Natural Language Processing (NLP) is a branch of Data Science which deals with Text data. Apart from numerical data, Text data is available to a great extent which is used to analyze and solve business problems. But before using the data for analysis or prediction, processing the data is important.

To prepare the text data for the model building we perform text preprocessing. It is the very first step of NLP projects. Some of the preprocessing steps are:

* Removing punctuations like . , ! $( ) * % @
* Removing URLs
* Removing Stop words
* Lower casing
* Tokenization
* Stemming
* Lemmatization


### SMS Spam Data for Text Preprocessing

In [1]:
import pandas as pd
data = pd.read_csv("spam.csv",encoding="ISO-8859-1")
print(data.head())

     v1                                                 v2 Unnamed: 2  \
0   ham  Go until jurong point, crazy.. Available only ...        NaN   
1   ham                      Ok lar... Joking wif u oni...        NaN   
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...        NaN   
3   ham  U dun say so early hor... U c already then say...        NaN   
4   ham  Nah I don't think he goes to usf, he lives aro...        NaN   

  Unnamed: 3 Unnamed: 4  
0        NaN        NaN  
1        NaN        NaN  
2        NaN        NaN  
3        NaN        NaN  
4        NaN        NaN  


In [4]:
#expanding the dispay of text sms column
pd.set_option('display.max_colwidth', -1)
#using only v1 and v2 column
data= data [['v1','v2']]
data.head()
data.shape

C:\Users\shesmeg\AppData\Local\Temp\ipykernel_106796\3969463988.py:2: FutureWarning: Passing a negative integer is deprecated in version 1.0 and will not be supported in future version. Instead, use None to not limit the column width.
  pd.set_option('display.max_colwidth', -1)


(5572, 2)

The data has 5572 rows and 2 columns. Let’s check the dependent variable distribution between spam and ham.

In [5]:
#checking the count of the dependent variable
data['v1'].value_counts()

ham     4825
spam    747 
Name: v1, dtype: int64

### Steps to Clean the Data

#### Punctuation Removal

In this step, all the punctuations from the text are removed. string library of Python contains some pre-defined list of punctuations such as ‘!”#$%&'()*+,-./:;?@[\]^_`{|}~’

In [8]:
#library that contains punctuation

import string
string.punctuation

#defining the function to remove punctuation
def remove_punctuation(text):
    punctuationfree="".join([i for i in text if i not in string.punctuation])
    return punctuationfree

#storing the puntuation free text

data['clean_msg']= data['v2'].apply(lambda x:remove_punctuation(x))
data

v1  \
0     ham    
1     ham    
2     spam   
3     ham    
4     ham    
...   ...    
5567  spam   
5568  ham    
5569  ham    
5570  ham    
5571  ham    

                                                                                                                                                                     v2  \
0     Go until jurong point, crazy.. Available only in bugis n great world la e buffet... Cine there got amore wat...                                                     
1     Ok lar... Joking wif u oni...                                                                                                                                       
2     Free entry in 2 a wkly comp to win FA Cup final tkts 21st May 2005. Text FA to 87121 to receive entry question(std txt rate)T&C's apply 08452810075over18's         
3     U dun say so early hor... U c already then say...                                                                                                                   
4     Nah I don't think he goes to usf, he lives around here though                                                                                                       
...                                                             ...                                                                                                       
5567  This is the 2nd time we have tried 2 contact u. U have won the å£750 Pound prize. 2 claim is easy, call 087187272008 NOW1! Only 10p per minute. BT-national-rate.   
5568  Will Ì_ b going to esplanade fr home?                                                                                                                               
5569  Pity, * was in mood for that. So...any other suggestions?                                                                                                           
5570  The guy did some bitching but I acted like i'd be interested in buying something else next week and he gave it to us for free                                       
5571  Rofl. Its true to its name                                                                                                                                          

                                                                                                                                                      clean_msg  
0     Go until jurong point crazy Available only in bugis n great world la e buffet Cine there got amore wat                                                     
1     Ok lar Joking wif u oni                                                                                                                                    
2     Free entry in 2 a wkly comp to win FA Cup final tkts 21st May 2005 Text FA to 87121 to receive entry questionstd txt rateTCs apply 08452810075over18s      
3     U dun say so early hor U c already then say                                                                                                                
4     Nah I dont think he goes to usf he lives around here though                                                                                                
...                                                           ...                                                                                                
5567  This is the 2nd time we have tried 2 contact u U have won the å£750 Pound prize 2 claim is easy call 087187272008 NOW1 Only 10p per minute BTnationalrate  
5568  Will Ì b going to esplanade fr home                                                                                                                        
5569  Pity  was in mood for that Soany other suggestions                                                                                                         
5570  The guy did some bitching but I acted like id be interested in buying something else next week and he gave it to us for free                               
5571

### Lowering the Text

It is one of the most common text preprocessing Python steps where the text is converted into the same case preferably lower case. But it is not necessary to do this step every time you are working on an NLP problem as for some problems lower casing can lead to loss of information.

For example, if in any project we are dealing with the emotions of a person, then the words written in upper cases can be a sign of frustration or excitement.

In [9]:
data['msg_lower']= data['clean_msg'].apply(lambda x: x.lower())

In [11]:
data

v1  \
0     ham    
1     ham    
2     spam   
3     ham    
4     ham    
...   ...    
5567  spam   
5568  ham    
5569  ham    
5570  ham    
5571  ham    

                                                                                                                                                                     v2  \
0     Go until jurong point, crazy.. Available only in bugis n great world la e buffet... Cine there got amore wat...                                                     
1     Ok lar... Joking wif u oni...                                                                                                                                       
2     Free entry in 2 a wkly comp to win FA Cup final tkts 21st May 2005. Text FA to 87121 to receive entry question(std txt rate)T&C's apply 08452810075over18's         
3     U dun say so early hor... U c already then say...                                                                                                                   
4     Nah I don't think he goes to usf, he lives around here though                                                                                                       
...                                                             ...                                                                                                       
5567  This is the 2nd time we have tried 2 contact u. U have won the å£750 Pound prize. 2 claim is easy, call 087187272008 NOW1! Only 10p per minute. BT-national-rate.   
5568  Will Ì_ b going to esplanade fr home?                                                                                                                               
5569  Pity, * was in mood for that. So...any other suggestions?                                                                                                           
5570  The guy did some bitching but I acted like i'd be interested in buying something else next week and he gave it to us for free                                       
5571  Rofl. Its true to its name                                                                                                                                          

                                                                                                                                                      clean_msg  \
0     Go until jurong point crazy Available only in bugis n great world la e buffet Cine there got amore wat                                                      
1     Ok lar Joking wif u oni                                                                                                                                     
2     Free entry in 2 a wkly comp to win FA Cup final tkts 21st May 2005 Text FA to 87121 to receive entry questionstd txt rateTCs apply 08452810075over18s       
3     U dun say so early hor U c already then say                                                                                                                 
4     Nah I dont think he goes to usf he lives around here though                                                                                                 
...                                                           ...                                                                                                 
5567  This is the 2nd time we have tried 2 contact u U have won the å£750 Pound prize 2 claim is easy call 087187272008 NOW1 Only 10p per minute BTnationalrate   
5568  Will Ì b going to esplanade fr home                                                                                                                         
5569  Pity  was in mood for that Soany other suggestions                                                                                                          
5570  The guy did some bitching but I acted like id be interested in buying something else next week and he gave it to us for free                          

### Tokenization

In this step, the text is split into smaller units. We can use either sentence tokenization or word tokenization based on our problem statement.

In [32]:
import nltk
nltk.download('punkt')
nltk.download('stopwords')
nltk.download('wordnet')
nltk.download('omw-1.4')

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\shesmeg\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\shesmeg\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\shesmeg\AppData\Roaming\nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package omw-1.4 to
[nltk_data]     C:\Users\shesmeg\AppData\Roaming\nltk_data...


True

In [19]:
#defining function for tokenization

from nltk.tokenize import word_tokenize

#applying function to the column
data['msg_tokenied']= data['msg_lower'].apply(lambda x: word_tokenize(x))

In [20]:
data

v1  \
0     ham    
1     ham    
2     spam   
3     ham    
4     ham    
...   ...    
5567  spam   
5568  ham    
5569  ham    
5570  ham    
5571  ham    

                                                                                                                                                                     v2  \
0     Go until jurong point, crazy.. Available only in bugis n great world la e buffet... Cine there got amore wat...                                                     
1     Ok lar... Joking wif u oni...                                                                                                                                       
2     Free entry in 2 a wkly comp to win FA Cup final tkts 21st May 2005. Text FA to 87121 to receive entry question(std txt rate)T&C's apply 08452810075over18's         
3     U dun say so early hor... U c already then say...                                                                                                                   
4     Nah I don't think he goes to usf, he lives around here though                                                                                                       
...                                                             ...                                                                                                       
5567  This is the 2nd time we have tried 2 contact u. U have won the å£750 Pound prize. 2 claim is easy, call 087187272008 NOW1! Only 10p per minute. BT-national-rate.   
5568  Will Ì_ b going to esplanade fr home?                                                                                                                               
5569  Pity, * was in mood for that. So...any other suggestions?                                                                                                           
5570  The guy did some bitching but I acted like i'd be interested in buying something else next week and he gave it to us for free                                       
5571  Rofl. Its true to its name                                                                                                                                          

                                                                                                                                                      clean_msg  \
0     Go until jurong point crazy Available only in bugis n great world la e buffet Cine there got amore wat                                                      
1     Ok lar Joking wif u oni                                                                                                                                     
2     Free entry in 2 a wkly comp to win FA Cup final tkts 21st May 2005 Text FA to 87121 to receive entry questionstd txt rateTCs apply 08452810075over18s       
3     U dun say so early hor U c already then say                                                                                                                 
4     Nah I dont think he goes to usf he lives around here though                                                                                                 
...                                                           ...                                                                                                 
5567  This is the 2nd time we have tried 2 contact u U have won the å£750 Pound prize 2 claim is easy call 087187272008 NOW1 Only 10p per minute BTnationalrate   
5568  Will Ì b going to esplanade fr home                                                                                                                         
5569  Pity  was in mood for that Soany other suggestions                                                                                                          
5570  The guy did some bitching but I acted like id be interested in buying something else next week and he gave it to us for free                          

### Stop Word Removal

Stopwords are the commonly used words and are removed from the text as they do not add any value to the analysis. These words carry less or no meaning.

NLTK library consists of a list of words that are considered stopwords for the English language. Some of them are : [i, me, my, myself, we, our, ours, ourselves, you, you’re, you’ve, you’ll, you’d, your, yours, yourself, yourselves, he, most, other, some, such, no, nor, not, only, own, same, so, then, too, very, s, t, can, will, just, don, don’t, should, should’ve, now, d, ll, m, o, re, ve, y, ain, aren’t, could, couldn’t, didn’t, didn’t]

But it is not necessary to use the provided list as stopwords as they should be chosen wisely based on the project. For example, ‘How’ can be a stop word for a model but can be important for some other problem where we are working on the queries of the customers. We can create a customized list of stop words for different problems.

In [23]:
#Stop words present in the library

stopwords = nltk.corpus.stopwords.words('english')
stopwords[0:10]
['i', 'me', 'my', 'myself', 'we', 'our', 'ours', 'ourselves', 'you', "you're"]

#defining the function to remove stopwords from tokenized text
def remove_stopwords(text):
    output= [i for i in text if i not in stopwords]
    return output

#applying the function
data['no_stopwords']= data['msg_tokenied'].apply(lambda x:remove_stopwords(x))

In [24]:
data

v1  \
0     ham    
1     ham    
2     spam   
3     ham    
4     ham    
...   ...    
5567  spam   
5568  ham    
5569  ham    
5570  ham    
5571  ham    

                                                                                                                                                                     v2  \
0     Go until jurong point, crazy.. Available only in bugis n great world la e buffet... Cine there got amore wat...                                                     
1     Ok lar... Joking wif u oni...                                                                                                                                       
2     Free entry in 2 a wkly comp to win FA Cup final tkts 21st May 2005. Text FA to 87121 to receive entry question(std txt rate)T&C's apply 08452810075over18's         
3     U dun say so early hor... U c already then say...                                                                                                                   
4     Nah I don't think he goes to usf, he lives around here though                                                                                                       
...                                                             ...                                                                                                       
5567  This is the 2nd time we have tried 2 contact u. U have won the å£750 Pound prize. 2 claim is easy, call 087187272008 NOW1! Only 10p per minute. BT-national-rate.   
5568  Will Ì_ b going to esplanade fr home?                                                                                                                               
5569  Pity, * was in mood for that. So...any other suggestions?                                                                                                           
5570  The guy did some bitching but I acted like i'd be interested in buying something else next week and he gave it to us for free                                       
5571  Rofl. Its true to its name                                                                                                                                          

                                                                                                                                                      clean_msg  \
0     Go until jurong point crazy Available only in bugis n great world la e buffet Cine there got amore wat                                                      
1     Ok lar Joking wif u oni                                                                                                                                     
2     Free entry in 2 a wkly comp to win FA Cup final tkts 21st May 2005 Text FA to 87121 to receive entry questionstd txt rateTCs apply 08452810075over18s       
3     U dun say so early hor U c already then say                                                                                                                 
4     Nah I dont think he goes to usf he lives around here though                                                                                                 
...                                                           ...                                                                                                 
5567  This is the 2nd time we have tried 2 contact u U have won the å£750 Pound prize 2 claim is easy call 087187272008 NOW1 Only 10p per minute BTnationalrate   
5568  Will Ì b going to esplanade fr home                                                                                                                         
5569  Pity  was in mood for that Soany other suggestions                                                                                                          
5570  The guy did some bitching but I acted like id be interested in buying something else next week and he gave it to us for free                          

### Stemming

It is also known as the text standardization step where the words are stemmed or diminished to their root/base form.  For example, words like ‘programmer’, ‘programming, ‘program’ will be stemmed to ‘program’.

But the disadvantage of stemming is that it stems the words such that its root form loses the meaning or it is not diminished to a proper English word. We will see this in the steps done below.

In [27]:
#importing the Stemming function from nltk library
from nltk.stem.porter import PorterStemmer
#defining the object for stemming
porter_stemmer = PorterStemmer()

#defining a function for stemming
def stemming(text):
    stem_text = [porter_stemmer.stem(word) for word in text]
    return stem_text

data['msg_stemmed']=data['no_stopwords'].apply(lambda x: stemming(x))

In [28]:
data

v1  \
0     ham    
1     ham    
2     spam   
3     ham    
4     ham    
...   ...    
5567  spam   
5568  ham    
5569  ham    
5570  ham    
5571  ham    

                                                                                                                                                                     v2  \
0     Go until jurong point, crazy.. Available only in bugis n great world la e buffet... Cine there got amore wat...                                                     
1     Ok lar... Joking wif u oni...                                                                                                                                       
2     Free entry in 2 a wkly comp to win FA Cup final tkts 21st May 2005. Text FA to 87121 to receive entry question(std txt rate)T&C's apply 08452810075over18's         
3     U dun say so early hor... U c already then say...                                                                                                                   
4     Nah I don't think he goes to usf, he lives around here though                                                                                                       
...                                                             ...                                                                                                       
5567  This is the 2nd time we have tried 2 contact u. U have won the å£750 Pound prize. 2 claim is easy, call 087187272008 NOW1! Only 10p per minute. BT-national-rate.   
5568  Will Ì_ b going to esplanade fr home?                                                                                                                               
5569  Pity, * was in mood for that. So...any other suggestions?                                                                                                           
5570  The guy did some bitching but I acted like i'd be interested in buying something else next week and he gave it to us for free                                       
5571  Rofl. Its true to its name                                                                                                                                          

                                                                                                                                                      clean_msg  \
0     Go until jurong point crazy Available only in bugis n great world la e buffet Cine there got amore wat                                                      
1     Ok lar Joking wif u oni                                                                                                                                     
2     Free entry in 2 a wkly comp to win FA Cup final tkts 21st May 2005 Text FA to 87121 to receive entry questionstd txt rateTCs apply 08452810075over18s       
3     U dun say so early hor U c already then say                                                                                                                 
4     Nah I dont think he goes to usf he lives around here though                                                                                                 
...                                                           ...                                                                                                 
5567  This is the 2nd time we have tried 2 contact u U have won the å£750 Pound prize 2 claim is easy call 087187272008 NOW1 Only 10p per minute BTnationalrate   
5568  Will Ì b going to esplanade fr home                                                                                                                         
5569  Pity  was in mood for that Soany other suggestions                                                                                                          
5570  The guy did some bitching but I acted like id be interested in buying something else next week and he gave it to us for free                          

Now let’s see how Lemmatization is different from Stemming.

### Lemmatization

It stems the word but makes sure that it does not lose its meaning.  Lemmatization has a pre-defined dictionary that stores the context of words and checks the word in the dictionary while diminishing.

The difference between Stemming and Lemmatization can be understood with the example provided below.

Original Word    After Stemming    After Lemmatization

goose            goos              goose\
geese            gees              goose

In [33]:
from nltk.stem import WordNetLemmatizer

#defining the object for Lemmatization
wordnet_lemmatizer = WordNetLemmatizer()

#defining the function for lemmatization
def lemmatizer(text):
    lemm_text = [wordnet_lemmatizer.lemmatize(word) for word in text]
    return lemm_text

data['msg_lemmatized']=data['no_stopwords'].apply(lambda x:lemmatizer(x))

In [34]:
data

v1  \
0     ham    
1     ham    
2     spam   
3     ham    
4     ham    
...   ...    
5567  spam   
5568  ham    
5569  ham    
5570  ham    
5571  ham    

                                                                                                                                                                     v2  \
0     Go until jurong point, crazy.. Available only in bugis n great world la e buffet... Cine there got amore wat...                                                     
1     Ok lar... Joking wif u oni...                                                                                                                                       
2     Free entry in 2 a wkly comp to win FA Cup final tkts 21st May 2005. Text FA to 87121 to receive entry question(std txt rate)T&C's apply 08452810075over18's         
3     U dun say so early hor... U c already then say...                                                                                                                   
4     Nah I don't think he goes to usf, he lives around here though                                                                                                       
...                                                             ...                                                                                                       
5567  This is the 2nd time we have tried 2 contact u. U have won the å£750 Pound prize. 2 claim is easy, call 087187272008 NOW1! Only 10p per minute. BT-national-rate.   
5568  Will Ì_ b going to esplanade fr home?                                                                                                                               
5569  Pity, * was in mood for that. So...any other suggestions?                                                                                                           
5570  The guy did some bitching but I acted like i'd be interested in buying something else next week and he gave it to us for free                                       
5571  Rofl. Its true to its name                                                                                                                                          

                                                                                                                                                      clean_msg  \
0     Go until jurong point crazy Available only in bugis n great world la e buffet Cine there got amore wat                                                      
1     Ok lar Joking wif u oni                                                                                                                                     
2     Free entry in 2 a wkly comp to win FA Cup final tkts 21st May 2005 Text FA to 87121 to receive entry questionstd txt rateTCs apply 08452810075over18s       
3     U dun say so early hor U c already then say                                                                                                                 
4     Nah I dont think he goes to usf he lives around here though                                                                                                 
...                                                           ...                                                                                                 
5567  This is the 2nd time we have tried 2 contact u U have won the å£750 Pound prize 2 claim is easy call 087187272008 NOW1 Only 10p per minute BTnationalrate   
5568  Will Ì b going to esplanade fr home                                                                                                                         
5569  Pity  was in mood for that Soany other suggestions                                                                                                          
5570  The guy did some bitching but I acted like id be interested in buying something else next week and he gave it to us for free                          

Output: The difference between Stemming and Lemmatization can be seen in the below output.

In the first row- crazy has been changed to crazi which has no meaning but for lemmatization, it remained the same i.e crazy

In the last row- goes has changed to goe while stemming but for lemmatization, it has converted into go which is meaningful.

### Endnotes

Apart from the steps shown in this article, many other steps are a part of preprocessing. Some of them are URL removal, HTML tags removal, Rare words removal, Frequent words removal, Spelling checking, and many more. The steps must be chosen based on the dataset you are working on and what is necessary for the project.